Improve sharing of skare3_tools data between users

javierggt opened the pull request:
## Description

The changes in this PR allow the `kadi` user to take over weekly processing. These changes include:
- moving the skare3_tools data into `$SKA/data`.
- making sure the caching decorators do not fail if the user has no write access to skare3_tools data.
- removing the option of having skare3_tools data in `~/skare3` (Linux and OSX) or ` %LOCALAPPDATA%\\skare3` (Windows). I decided we will never use it. We will always use `$SKA/data/`, and if we really need to change that, we can define `SKARE3_TOOLS_DATA`.
- added a few book-keeping functions for storing test results.
- replacing some uses of `os.path` by `pathlib`.

Please don't pay attention to linter. I made another PR that will come after this and fix linting (#108).

## Interface impacts
<!-- API changes, file format updates, coordination of changes with the community. -->

After this PR any user will be able to use the skare3_tools installed on HEAD to get the package information. All data will be stored in `$SKA/data/skare3` or in `$SKARE3_TOOLS_DATA`. This includes test result data as well as cached package data.

If the cache is old, and the user does not have write access to the cache directory, the old information in the cache is returned, unless `SKARE3_REPO_INFO_LATEST=true`, in which case the latest info is returned but the cache is not updated.


## Testing
<!-- If relevant describe any special setup for testing. -->

### Unit tests
<!-- At least one of these must be checked if unit tests exist. DELETE the unchecked/untested options. -->
- [x] No unit tests


### Functional tests
This branch is installed by hand in the ska3-masters environment owned by kadi and the dashboard and testing has used it.

Anyone should be able to do the following and get reasonable-looking result:
```
from skare3_tools import packages, config, test_results

assert config.CONFIG['data_dir'] == f'{os.environ["SKA"]}/data/skare3/skare3_data/data'

pkgs = packages.get_package_list()
pkgs_info = packages.get_repositories_info()
tr = test_results.get(stream='ska3-masters')
```

On HEAD, it should use the data that the kadi user has put in `/proj/sot/ska`. If you want to use your own location, you can set the `SKARE3_TOOLS_DATA` environment variable, and in this case
```
config.CONFIG['data_dir'] == os.path.join(os.environ["SKARE3_TOOLS_DATA"], 'data')
```

javierggt committed:
packages.py: do not update cache if there is no write access

javierggt committed:
config.py: add $SKA/data/skare3/skare3_data as one of the options for data storage

javierggt committed:
test_results.py: added streams, remove_older_than and remove methods. Replaced uses of os.path with pathlib. When adding test results, copy to a temp dir first to check, and verify that destination does not exist already.

javierggt committed:
add symlink to latest in stream

javierggt committed:
remove possibility to store in ~/.skare3 or LOCALAPPDATA%\\skare3

jeanconn commented:
Regarding "Anyone should be able to do the following and get reasonable-looking result:" that seemed like the functional tests that I should double-check run as myself (jeanconn) because this shouldn't write back into skare3_tools data.  But I need to go initialize some credentials:
```
WARNING:root:Failed to get info on acisops/RT_Web_Pages: GithubAPI authentication credentials are not initialized
...
```

javierggt committed:
skip fetching repo info if cache cannot be written to

javierggt commented:
@jeanconn 

I fixed your issue by skipping updates if one does not have write access to the cache directory. I also added a way around it by defining the `SKARE3_REPO_INFO_LATEST=true` environment variable in case one wants to get the latest info and is not reflected in the cache owned by someone else.

javierggt committed:
change env var name

jeanconn commented on skare3_tools/config.py:
```diff
@@ -104,16 +101,11 @@
-    elif os.path.exists(home_dir) and os.access(home_dir, os.W_OK):
-        # can use this in linux and Mac OS
-        app_data_dir = os.path.join(home_dir, ".skare3")
+    elif os.path.exists(ska_data_dir):
```
For non-head use, it looks like this will fail the first time with "Exception: Could not figure out where to place skare3_tools configuration" unless the user makes a skare3_data directory or has manually set SKARE3_TOOLS_DATA.

jeanconn replied:
That's probably fine but not sure if it needs a hint someplace in docs that I missed.

javierggt replied:
Yes, originally I had it so it would create it in the home directory, but in reality that just needs to be set beforehand. Nothing will be created automatically.

javierggt replied:
I changed the error message, so it tells you what to do.

jeanconn replied:
Thanks!  Though now I'm not sure if the functional test has a hiccup or the config location isn't perfect -- with or without /data?
```
Exception: Could not figure out the location of the skare3_tools configuration.
Either create the $SKA/data/skare3/skare3_data directory
or set the SKARE3_TOOLS_DATA environmental variable.

In [2]: exit
(ska3-masters) flame:skare3_tools jean$ mkdir $SKA/data/skare3/skare3_data
(ska3-masters) flame:skare3_tools jean$ ipython
Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:51:20) [Clang 16.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.22.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from skare3_tools import packages, config, test_results

In [2]: import os

In [3]: assert config.CONFIG['data_dir'] == f'{os.environ["SKA"]}/data/skare3/skare3_data'
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[3], line 1
----> 1 assert config.CONFIG['data_dir'] == f'{os.environ["SKA"]}/data/skare3/skare3_data'

AssertionError: 

In [4]: %debug
> <ipython-input-3-d735df94dd35>(1)<module>()
----> 1 assert config.CONFIG['data_dir'] == f'{os.environ["SKA"]}/data/skare3/skare3_data'

ipdb> config.CONFIG['data_dir']
'/Users/jean/ska/data/skare3/skare3_data/data'
```

javierggt replied:
and what is SKA? I'm guessing there is a symlink somewhere, and I need to follow it.

jeanconn replied:
I think there's no symlink involved here on my mac.

jeanconn replied:
```
(ska3-masters) flame:skare3_tools jean$ ipython
Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:51:20) [Clang 16.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.22.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import os

In [2]: os.environ['SKA']
Out[2]: '/Users/jean/ska'

In [3]: from skare3_tools import packages, config, test_results
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 from skare3_tools import packages, config, test_results

File ~/git/skare3_tools/skare3_tools/packages.py:89
     87 from packaging.version import Version, InvalidVersion
     88 from skare3_tools import github
---> 89 from skare3_tools.config import CONFIG
     90 from pathlib import Path
     92 class NetworkException(Exception):

File ~/git/skare3_tools/skare3_tools/config.py:153
    150 # this could be replaced by a lazy attribute in shiny
    151 # (https://www.python.org/dev/peps/pep-0562/)
    152 CONFIG = _DEFAULT_CONFIG.copy()
--> 153 init()

File ~/git/skare3_tools/skare3_tools/config.py:126, in init(config, reset)
    124 app_data_dir = _app_data_dir_()
    125 if app_data_dir is None:
--> 126     raise Exception(
    127         "Could not figure out the location of the skare3_tools configuration.\n"
    128         "Either create the $SKA/data/skare3/skare3_data directory\n"
    129         "or set the SKARE3_TOOLS_DATA environmental variable."
    130     )
    131 config_file = os.path.join(app_data_dir, "config.json")
    132 exists = os.path.exists(config_file)

Exception: Could not figure out the location of the skare3_tools configuration.
Either create the $SKA/data/skare3/skare3_data directory
or set the SKARE3_TOOLS_DATA environmental variable.

In [4]: exit
(ska3-masters) flame:skare3_tools jean$ mkdir -p $SKA/data/skare3/skare3_data
(ska3-masters) flame:skare3_tools jean$ ipython
Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:51:20) [Clang 16.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.22.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import os

In [2]: os.environ['SKA']
Out[2]: '/Users/jean/ska'

In [3]: from skare3_tools import packages, config, test_results

In [4]: config.CONFIG['data_dir']
Out[4]: '/Users/jean/ska/data/skare3/skare3_data/data'

In [5]: assert config.CONFIG['data_dir'] == f'{os.environ["SKA"]}/data/skare3/skare3_data'
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 assert config.CONFIG['data_dir'] == f'{os.environ["SKA"]}/data/skare3/skare3_data'

AssertionError: 
```

javierggt replied:
ok, can you show me what is `f'{os.environ["SKA"]}/data/skare3/skare3_data'`?

jeanconn replied:
Right, there seems to be this ambiguity about /data .

```
In [1]: import os

In [2]: from skare3_tools import packages, config, test_results

In [3]: f'{os.environ["SKA"]}/data/skare3/skare3_data'
Out[3]: '/Users/jean/ska/data/skare3/skare3_data'

In [4]: config.CONFIG['data_dir']
Out[4]: '/Users/jean/ska/data/skare3/skare3_data/data'
```

javierggt replied:
ah!, I see now. It's the test here in the PR.

jeanconn replied:
Right, it is just the functional "assert" test in the PR.  But the weird thing for me is that assert is fine on HEAD and doesn't work (mismatch on /data) on my mac.

javierggt replied:
that was because on HEAD I copied things by hand from my directory to kadi's (I believe) so that one was inconsistent. I moved it so now it agrees with this convention.

javierggt committed:
change message in case SKARE3_TOOLS_DATA is undefined
